omaha/protocol.go

```diff
@@ -55,18 +56,71 @@ func NewRequest() *Request {
 	}
 }
 
-func (r *Request) AddApp(id, version string) *App {
-	a := &App{Id: id, Version: version}
+func (r *Request) AddApp(id, version string) *AppRequest {
+	a := &AppRequest{Id: id, Version: version}
 	r.Apps = append(r.Apps, a)
 	return a
 }
 
+type AppRequest struct {
+	Ping        *PingRequest    `xml:"ping"`
+	UpdateCheck *UpdateRequest  `xml:"updatecheck"`
+	Events      []*EventRequest `xml:"event" json:",omitempty"`
+	Id          string          `xml:"appid,attr,omitempty"`
+	Version     string          `xml:"version,attr,omitempty"`
+	NextVersion string          `xml:"nextversion,attr,omitempty"`
+	Lang        string          `xml:"lang,attr,omitempty"`
+	Client      string          `xml:"client,attr,omitempty"`
+	InstallAge  string          `xml:"installage,attr,omitempty"`
+
+	// update engine extensions
+	Track     string `xml:"track,attr,omitempty"`
+	FromTrack string `xml:"from_track,attr,omitempty"`
+
+	// coreos update engine extensions
+	BootId    string `xml:"bootid,attr,omitempty"`
+	MachineID string `xml:"machineid,attr,omitempty"`
+	OEM       string `xml:"oem,attr,omitempty"`
+}
+
+func (a *AppRequest) AddUpdateCheck() *UpdateRequest {
+	a.UpdateCheck = &UpdateRequest{}
+	return a.UpdateCheck
+}
+
+func (a *AppRequest) AddPing() *PingRequest {
+	a.Ping = new(PingRequest)
```

Comment:
That gets fixed later on in the series, at this point in the patch series this is moving existing code which uses the new() operator.